src/signal-handling.c

```diff
@@ -88,12 +93,134 @@ JL_DLLEXPORT void jl_exit_on_sigint(int on)
     exit_on_sigint = on;
 }
 
+void jl_show_sigill(void *_ctx);
+
 #if defined(_WIN32)
 #include "signals-win.c"
 #else
 #include "signals-unix.c"
 #endif
 
+void jl_show_sigill(void *_ctx)
+{
+#if defined(_CPU_X86_64_) || defined(_CPU_X86_)
+    uintptr_t pc = 0;
+#  if defined(_OS_LINUX_) && defined(_CPU_X86_64_)
+    pc = ((ucontext_t*)_ctx)->uc_mcontext.gregs[REG_RIP];
+#  elif defined(_OS_FREEBSD_) && defined(_CPU_X86_64_)
+    pc = ((ucontext_t*)_ctx)->uc_mcontext.mc_rip;
+#  elif defined(_OS_LINUX_) && defined(_CPU_X86_)
+    pc = ((ucontext_t*)_ctx)->uc_mcontext.gregs[REG_EIP];
+#  elif defined(_OS_FREEBSD_) && defined(_CPU_X86_)
+    pc = ((ucontext_t*)_ctx)->uc_mcontext.mc_eip;
+#  elif defined(_OS_DARWIN_)
+    pc = ((ucontext64_t*)_ctx)->uc_mcontext64->__ss.__rip;
+#  elif defined(_OS_WINDOWS_) && defined(_CPU_X86_)
+    pc = ((CONTEXT*)_ctx)->Eip;
+#  elif defined(_OS_WINDOWS_) && defined(_CPU_X86_64_)
+    pc = ((CONTEXT*)_ctx)->Rip;
+#  endif
+    // unsupported platform
+    if (!pc)
+        return;
+    uint8_t inst[15]; // max length of x86 instruction
+    int len = sizeof(inst);
+    // since we got a SIGILL and not a SIGSEGV or SIGBUS assume that
+    // the `pc` is pointed to valid memory.
+    // However, this does not mean that `pc + 14` is valid memory.
+    uintptr_t page_start = pc / jl_page_size * jl_page_size;
+    uintptr_t next_page = page_start + jl_page_size;
+    if (next_page < pc + len) {
+        int valid = 0;
+        // The max instruction length crosses page boundary.
+        // Check if it's safe to read the next page
+#ifdef _OS_WINDOWS_
+        MEMORY_BASIC_INFORMATION mbi;
+        if (VirtualQuery((void*)next_page, &mbi, sizeof(mbi))) {
+            if (mbi.State == MEM_COMMIT) {
+                // If the address is not both executable and readable,
+                // it'll not hold any part of the instruction.
+                if (mbi.Protect == PAGE_EXECUTE || mbi.Protect == PAGE_EXECUTE_READ ||
+                    mbi.Protect == PAGE_EXECUTE_READWRITE ||
+                    mbi.Protect == PAGE_EXECUTE_WRITECOPY) {
+                    valid = 1;
+                }
+            }
+        }
+#else
+#  if defined(_OS_FREEBSD_) || defined(_OS_DARWIN_)
+        char mvec;
+#  else
+        unsigned char mvec;
+#  endif
+        valid = mincore((void*)next_page, jl_page_size, &mvec) != -1;
```

Comment:
I've found that ```if (msync((void*)(fault_addr & ~(jl_page_size - 1)), 1, MS_ASYNC) != 0)``` is more accurate (tests whether the kernel has defined a memory region for that virtual address, rather than testing if that address currently resides in physical memory)